src/tests/ftest/util/general_utils.py

```diff
@@ -605,19 +605,23 @@ def stop_processes(hosts, pattern, verbose=True, timeout=60):
     result = {}
     log = getLogger()
     log.info("Killing any processes on %s that match: %s", hosts, pattern)
+
+    # Filter out zombie (defunct) processes from ps output
+    ps_with_zombie_filter = r"ps x | grep -E {} | grep -vE '\<(grep|defunct)\>'"
+
     if hosts is not None:
         commands = [
             "rc=0",
             "set -x",
             "ps aux | grep -E {}".format(pattern),
-            "if pgrep --list-full {}".format(pattern),
+            "if " + ps_with_zombie_filter.format(pattern),
```

Comment:
Did you consider the pgrep `--runstates` option:
```
       -r, --runstates D,R,S,Z,...
              Match only processes which match the process state.
```
Since we don't want to match defunct/zombie (Z) processes and probably can't do anything with uninterruptable sleep (D) processes, only searching for running (R) and interruptible sleep (S) process would make sense:
```suggestion
            "if pgrep --list-full --runstates R,S {}".format(pattern),
```

That said this is a general function, so this sort of feature should be optional, e.g.:
```
def stop_processes(hosts, pattern, verbose=True, timeout=60, run_states=None):
...
        run_states (str, optional): filter results by run state. Defaults to None.
            Options include the following; multiple options can be comma-separated:
                D    uninterruptible sleep (usually IO)
                R    running or runnable
                S    interruptible sleep
                Z    defunct ("zombie") process
...
            "if pgrep --list-full {}{}".format(
                "--runstates {} ".format(run_states) if run_states else "",
                pattern),
```

Reply:
> Did you consider the pgrep --runstates option:
> 
>        -r, --runstates D,R,S,Z,...
>               Match only processes which match the process state.
>
> Since we don't want to match defunct/zombie (Z) processes and probably can't do anything with uninterruptable sleep (D) processes, only searching for running (R) and interruptible sleep (S) process would make sense:

I'd prefer `--runstates` over `ps x ... | grep`, but CentOS `pgrep` doesn't seem to support it:

```
$ cat /etc/redhat-release
CentOS Linux release 7.8.2003 (Core)
$ pgrep --runstates R,S .
pgrep: unrecognized option '--runstates'
```

Reply:
> ... this is a general function, so this sort of feature should be optional, e.g.:

How about an `added_filter` option to keep the function general?

```diff

-def stop_processes(hosts, pattern, verbose=True, timeout=60):
+def stop_processes(hosts, pattern, verbose=True, timeout=60, added_filter=""):
     """Stop the processes on each hosts that match the pattern.

     Args:
@@ -607,21 +607,21 @@ def stop_processes(hosts, pattern, verbose=True, timeout=60):
     log.info("Killing any processes on %s that match: %s", hosts, pattern)

     # Filter out zombie (defunct) processes from ps output
-    ps_with_zombie_filter = r"ps x | grep -E {} | grep -vE '\<(grep|defunct)\>'"
+    ps_cmd = r"ps x | grep -E {} " + added_filter

     if hosts is not None:
         commands = [
             "rc=0",
             "set -x",
             "ps aux | grep -E {}".format(pattern),
-            "if " + ps_with_zombie_filter.format(pattern),
+            "if " + ps_cmd.format(pattern),
             "then rc=1",
             "sudo pkill {}".format(pattern),
             "sleep 5",
-            "if " + ps_with_zombie_filter.format(pattern),
+            "if " + ps_cmd.format(pattern),
             "then pkill --signal ABRT {}".format(pattern),
             "sleep 1",
-            "if " + ps_with_zombie_filter.format(pattern),
+            "if " + ps_cmd.format(pattern),
             "then pkill --signal KILL {}".format(pattern),
             "fi",
             "fi",
```

Reply:
> I'd prefer `--runstates` over `ps x ... | grep`, but CentOS `pgrep` doesn't seem to support it:
> 
> ```
> $ cat /etc/redhat-release
> CentOS Linux release 7.8.2003 (Core)
> $ pgrep --runstates R,S .
> pgrep: unrecognized option '--runstates'
> ...
> $ pgrep -V
> pgrep from procps-ng 3.3.10
> ```
Nuts! Apparently it was introduced in procps-ng-3.3.16 (https://gitlab.com/procps-ng/procps/-/blob/59c88e18f29000ceaf7e5f98181b07be443cf12f/NEWS).  I was hoping we could just make it a minimum version requirement for the daos-tests RPM, but I don't see a procps-ng-3.3.16 RPM available in our repos.

Reply:
> > ... this is a general function, so this sort of feature should be optional, e.g.:
> 
> How about an `added_filter` option to keep the function general?
> 

That could work.  I'm concerned that a `ps x | grep -E <pattern>` may end up matching something different than `pgrep <pattern>`.   As not to introduce any possible problems and to make it slightly easier to use (where the user doesn't need to include `" | grep -E "` with their filter), might I suggest the following modification:
```
        if added_filter:
            ps_cmd += "ps x | grep -E {} | grep -E {}".format(pattern, added_filter)
        else:
            pc_cmd = "pgrep --list-full {}".format(pattern)
```

Reply:
>  ... I suggest the following modification:
> 
> ```
> if added_filter:
>   ps_cmd = "ps x | grep -E {} | grep -E {}".format(pattern, added_filter)
> else:
>   pc_cmd = "pgrep --list-full {}".format(pattern)
> ```

Looks good to me! 👍

Reply:
I discussed the zombie `orterun` with @frostedcmos, and we're concerned one of these zombies might disrupt subsequent tests, e.g., if it's leaving around stale resources that get inadvertently used by other tests. Unfortunately, it looks like the [way to kill an already-dead process](https://www.linuxjournal.com/content/how-kill-zombie-processes-linux) is to kill it's parent.  I'm hoping `avocado` isn't the parent of the `orterun`, rather -- I hope it's `orterun`'s (great-)grandparent (e.g., it spawns a `sh` process which invokes `orterun`). I was going to run some diagnostics to investigate the process tree, and see if removing the zombie's parent would be possible.